pandas/indexes/multi.py

```diff
@@ -656,9 +665,74 @@ def _has_complex_internals(self):
         return True
 
     @cache_readonly
+    def is_monotonic(self):
+
+        def level_values(level):
+            unique = self.levels[level]
+            labels = self.labels[level]
+            return algos.take_1d(unique.values, labels,
+                                 fill_value=unique._na_value)
```

Comment:
I think the case where this could potentially fail currently is something like:

`pd.MultiIndex.from_product([[1.0, np.nan, 2.0], ['a', 'b', 'c']])`

You've made some interesting life choices if you've got NaNs in your index (or really if you're intentionally using a float index at all IMO), but I'd expect `is_monotonic` to be reliably `False` in this case, which I don't think my proposed implementation guarantees, since any sorting algorithm is going to return garbage if the inputs contain NaN.